From c69833f92f9fae6fe3cbd2b7862a45e5dc7cb917 Mon Sep 17 00:00:00 2001
From: "Thulasimani,Sivakumar" <sivakumar.thulasimani@intel.com>
Date: Mon, 21 Sep 2015 17:17:38 +0530
Subject: [PATCH] SQUASHME [VPG]: drm/i915: Selectively do upfront link
 training

Upfront link training is needed to handle USB Type C displays
which also requires overhead of retrying at multiple lanecounts.
This retry affects timings expected in automated compliance tests,
hence this patch performs upfront link training only when automated
compliance test suite is not running.

SQUASHME! - This patch should be squashed into the following existing patch:
Author:    Durgadoss R <durgadoss.r@intel.com>
Date:      Fri, 22 May 2015 14:30:07 +0530
Commit:    45cb921339e44f6c2c8ea7bf5b21f3d384769296
Change-Id: I2ae866b32a0998abcf32826efa2504e203dc8ffa
Subject:   FOR_UPSTREAM [VPG]: drm/i915: Enable Upfront link training
           for type-C DP support

Change-Id: Ib71457b321206776da2838f42c05d6bc17173c71
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-5971
Signed-off-by: Sivakumar Thulasimani <sivakumar.thulasimani@intel.com>
---
 drivers/gpu/drm/i915/intel_dp.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_dp.c b/drivers/gpu/drm/i915/intel_dp.c
index a7d0d58..9af1337 100644
--- a/drivers/gpu/drm/i915/intel_dp.c
+++ b/drivers/gpu/drm/i915/intel_dp.c
@@ -4860,6 +4860,25 @@ intel_dp_detect(struct drm_connector *connector, bool force)
 			DRM_DEBUG_DRIVER("CP or sink specific irq unhandled\n");
 	}
 
+	if (IS_CHERRYVIEW(dev) &&
+		intel_dp->compliance_test_type != DP_TEST_LINK_TRAINING &&
+			intel_encoder->type == INTEL_OUTPUT_DISPLAYPORT) {
+
+		/*
+		 * TODO: Need to test connected boot scenario once platform
+		 * patches are ready. This path is tested on reworked-RVP only.
+		 */
+		if (intel_encoder->connectors_active &&
+						crtc && crtc->enabled) {
+			intel_crtc = to_intel_crtc(crtc);
+			DRM_DEBUG_KMS("Disabling crtc %c for upfront LT\n",
+					pipe_name(intel_crtc->pipe));
+			intel_crtc_control(crtc, false);
+		}
+		chv_upfront_link_train(dev, intel_dp, intel_crtc);
+	}
+
+
 out:
 #ifdef CONFIG_SUPPORT_LPDMA_HDMI_AUDIO
 	if (IS_VALLEYVIEW(dev) && !is_edp(intel_dp) &&
-- 
1.9.1

